R/standardize_treatment.R

```diff
@@ -0,0 +1,445 @@
+                                  lowlim = 0, uplim = nrow(X0), scale_sample_size = F,
+                                  data_in = NULL, verbose = TRUE,
+                                  return_program = TRUE, exact_global = F,
+                                  init_uniform = F, eps_abs = 1e-5,
+                                  eps_rel = 1e-5, ...) {
+
+  # ensure that covariate matrices are matrices and get total number of units
+  X0 <- as.matrix(X0)
+  Xtau <- as.matrix(Xtau)
+  n <- nrow(X0)
+
+  # convert site indicators to factor, get index of sites, and compute # sites
+  S_factor <- as.factor(S)
+  unique_S <- levels(S_factor)
+  J <- length(unique_S)
+
+  # split covariate matrix by site and compute number of units in each site
+  X0s <- split.data.frame(X0, S_factor)
+  Xtaus <- split.data.frame(Xtau, S_factor)
+  nj <- sapply(X0s, nrow)
+
+  # get row IDs of units by site
+  idxs <- split(1:nrow(X0), S_factor)
+
+  # ensure that target is a vector
+  target <- c(target)
+
+  # check arguments for issues
+  check_args_treatment(X0, Xtau, target, S, Z, pscores, X0s, Xtaus,
+                       as.numeric(nj), lambda, lowlim, uplim, data_in)
+
+  # create propensity score multipliers and split by site
+  pro_trt <- Z / (pscores * nj[S_factor])
+  pro_trt_split <- split(pro_trt, S_factor)
+  pro_ctr <- (1 - Z) / ((1 - pscores) * nj[S_factor])
+  pro_ctr_split <- split(pro_ctr, S_factor)
+
+  # define number of auxiliary weights
+  aux_dim <- (J * ncol(X0)) + (J * ncol(Xtau))
+
+  # construct linear term vector
+  if(verbose) message("Creating linear term vector...")
+  if(is.null(data_in$q)) {
+    q <- create_q_vector_treatment(Xtaus, pro_trt_split, target, aux_dim)
+  } else {
+    q <- data_in$q
+  }
+
+  # construct quadratic term matrix
+  if(verbose) message("Creating quadratic term matrix...")
+  if(is.null(data_in$P)) {
+    P <- create_P_matrix_treatment(n, aux_dim)
+  } else {
+    P <- data_in$P
+  }
+  I0 <- create_I0_matrix_treatment(pro_trt_split, pro_ctr_split,
+                                   scale_sample_size, nj, n, aux_dim)
+  P <- P + lambda * I0
+
+  # construct constraint matrix
+  if(verbose) message("Creating constraint matrix...")
+  if(is.null(data_in$constraints)) {
+    constraints <- create_constraints_treatment(X0s, Xtaus, target, Z, S_factor,
+                                                pro_trt_split, pro_ctr_split,
+                                                lowlim, uplim, exact_global, verbose)
+  } else {
+    constraints <- data_in$constraints
+    constraints$l[(J + 1):(J + n)] <- lowlim
+    constraints$u[(J + 1):(J + n)] <- uplim
+  }
+
+  # set optimization settings
+  if(verbose) {
+    settings <- osqp::osqpSettings(verbose = TRUE, eps_abs = 1e-6,
+                                   eps_rel = 1e-6, max_iter = 5000)
+  } else {
+    settings <- osqp::osqpSettings(verbose = FALSE, eps_abs = 1e-6,
+                                   eps_rel = 1e-6, max_iter = 5000)
+  }
```

Comment:
I think this part is redundant?

Reply:
I think so too. I removed it. I also removed it from the `standardize.R` function, if that's OK with you.